src/datalink/bpf.rs

```diff
@@ -253,30 +261,43 @@ pub struct DataLinkChannelIteratorImpl<'a> {
 
 impl<'a> EthernetDataLinkChannelIterator<'a> for DataLinkChannelIteratorImpl<'a> {
     fn next(&mut self) -> io::Result<EthernetPacket> {
+        // Loopback packets arrive with a 4 byte header instead of normal ethernet header.
+        // Discard that header and replace with zeroed out ethernet header.
+        let (header_size, buffer_offset) = if self.pc.loopback {
+            (4, EthernetPacket::minimum_packet_size())
+        } else {
+            (0, 0)
+        };
         if self.packets.is_empty() {
+            let buffer = &mut self.pc.read_buffer[buffer_offset..];
             let buflen = match unsafe {
                 libc::read(self.pc.fd.fd,
-                           self.pc.read_buffer.as_ptr() as *mut libc::c_void,
-                           self.pc.read_buffer.len() as libc::size_t)
+                           buffer.as_ptr() as *mut libc::c_void,
+                           buffer.len() as libc::size_t)
             } {
                 len if len > 0 => len,
                 _ => return Err(io::Error::last_os_error()),
             };
-            let mut ptr = self.pc.read_buffer.as_mut_ptr();
-            let end = unsafe { self.pc.read_buffer.as_ptr().offset(buflen as isize) };
+            let mut ptr = buffer.as_mut_ptr();
+            let end = unsafe { buffer.as_ptr().offset(buflen as isize) };
             while (ptr as *const u8) < end {
                 unsafe {
                     let packet: *const bpf::bpf_hdr = mem::transmute(ptr);
                     let start = ptr as isize + (*packet).bh_hdrlen as isize -
-                                self.pc.read_buffer.as_ptr() as isize;
-                    self.packets.push_back((start as usize + self.pc.header_size,
-                                            (*packet).bh_caplen as usize - self.pc.header_size));
+                                buffer.as_ptr() as isize;
+                    self.packets.push_back((start as usize + header_size,
+                                            (*packet).bh_caplen as usize - header_size));
                     let offset = (*packet).bh_hdrlen as isize + (*packet).bh_caplen as isize;
                     ptr = ptr.offset(bpf::BPF_WORDALIGN(offset));
                 }
             }
         }
-        let (start, len) = self.packets.pop_front().unwrap();
+        let (start, mut len) = self.packets.pop_front().unwrap();
+        len += buffer_offset;
+        // Zero out part that will become fake ethernet header if on loopback.
+        for i in (&mut self.pc.read_buffer[start..start + buffer_offset]).iter_mut() {
+            *i = 0;
+        }
```

Comment:
Is there any shorter way to zero out a slice? It feels like there should be a nice functional style, one-liner laying around that I'm missing :)